dataprep/eda/eda_frame.py

```diff
@@ -0,0 +1,367 @@
+"""Defines DataArray."""
+
+from functools import reduce
+from math import ceil
+from typing import Any, List, Optional, Sequence, Tuple, Union, cast, Dict
+from collections import Counter
+import warnings
+
+import dask
+import dask.array as da
+import dask.dataframe as dd
+import numpy as np
+import pandas as pd
+import pandas._libs.missing as libmissing
+
+from .dtypes_v2 import (
+    NUMERICAL_DTYPES,
+    DType,
+    DTypeDef,
+    detect_dtype,
+    Nominal,
+)
+
+DataFrame = Union[pd.DataFrame, dd.DataFrame, "EDAFrame"]
+
+
+class EDAFrame:
+    """EDAFrame provides an abstraction over dask DataFrame
+    and dask Array. The reason is that sometimes some algorithms
+    only works on the Array and not the DataFrame. However,
+    the cost for getting the array from a dask DataFrame (with known length)
+    is non trivial. Instead of computing the array from a dask
+    DataFrame again and again, it would be better do that once.
+
+    Other reasons to have a separate EDAFrame abstraction includes
+    converting the column names to string without modifying the
+    DataFrame from user, and preprocessings like dropna and type detection.
+
+    Parameters
+    ----------
+    df
+        The DataFrame
+    value_length
+        Whether to compute the lengths of the array.
+        This triggers a read on the data thus expensive if the passed in df
+        is a dask DataFrame.
+        If a pandas DataFrame passed in, lengths will always be compute.
+    repartition
+        Whether to repartition the DataFrame into 128M chunks.
+    dtype: str or DType or dict of str or dict of DType, default None
+        Specify Data Types for designated column or all columns.
+        E.g.  dtype = {"a": Continuous, "b": "Nominal"} or
+        dtype = {"a": Continuous(), "b": "nominal"}
+        or dtype = Continuous() or dtype = "Continuous" or dtype = Continuous()
+    """
+
+    # pylint: disable=too-many-instance-attributes
+    _ddf: dd.DataFrame
+    _values: da.Array
+    _nulls: Union[da.Array, np.ndarray]
+    _columns: pd.Index
+    _eda_dtypes: Dict[str, DType] = {}
+    _str_col_cache: Dict[Tuple[str, bool], dd.Series] = {}
+    _nulls_cnt: Dict[str, int] = {}
+    _head: Optional[pd.DataFrame] = None
+    _shape: Optional[Tuple[int, int]] = None
+
+    # pylint: disable = too-many-branches
+    def __init__(
+        self,
+        df: Optional[DataFrame] = None,
+        value_length: bool = False,
+        repartition: bool = True,
+        dtype: Optional[DTypeDef] = None,
+    ) -> None:
+
+        _suppress_warnings()
+
+        if df is None:
```

Comment:
This is for supporting of empty edaframe like `pd.DataFrame()`. In `__getitem__` I will create an empty `EDAFrame` and then assign related attribute to avoid the cost of processing dataframe.